File: app/src/main/java/com/liweiyap/narradir/util/audio/CharacterDescriptionMediaPlayer.java
package com.liweiyap.narradir.util.audio;

import android.content.Context;
import android.media.MediaPlayer;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.liweiyap.narradir.R;

public class CharacterDescriptionMediaPlayer implements MediaPlayerController
{
    public CharacterDescriptionMediaPlayer(final @NonNull Context context)
    {
        mContext = context;
    }

    public void destroy()
    {
        mContext = null;
        free();
    }

    @Override
    public void play(final String res, final float volume, final MediaPlayer.OnCompletionListener listener)
    {
        if (mMediaPlayer != null)
        {
            mMediaPlayer.stop();
        }

        if (res == null)
        {
            return;
        }

        try
        {
            mMediaPlayer = create(res);
            if (mMediaPlayer == null)
            {
                return;
            }

            setVolume(volume);
            mMediaPlayer.setOnCompletionListener(listener);
            mMediaPlayer.start();
        }
        catch (IllegalStateException e)
        {
            e.printStackTrace();
        }
    }

    @Override
    public void resume()
    {
        if (mMediaPlayer == null)
        {
            return;
        }

        mMediaPlayer.seekTo(mMediaPlayerCurrentLength);
        mMediaPlayer.start();
    }

    @Override
    public void pause()
    {
        if (mMediaPlayer == null)
        {
            return;
        }

        mMediaPlayer.pause();
        mMediaPlayerCurrentLength = mMediaPlayer.getCurrentPosition();
    }

    @Override
    public void stop()
    {
        if (mMediaPlayer == null)
        {
            return;
        }

        mMediaPlayer.setOnCompletionListener(null);
        mMediaPlayer.stop();
    }

    @Override
    public void free()
    {
        if (mMediaPlayer == null)
        {
            return;
        }

        mMediaPlayer.release();
        mMediaPlayer = null;
    }

    @Override
    public void setVolume(final float volume)
    {
        if (mMediaPlayer == null)
        {
            return;
        }

        mMediaPlayer.setVolume(volume, volume);
    }

    private @Nullable MediaPlayer create(final @NonNull String res)
    {
        // does not look elegant but it's safe because RawRes ID integers are non-constant from Gradle version >4.
        // no need to call prepare(); create() does that for you (https://stackoverflow.com/a/59682667/12367873)
        if (res.equals(mContext.getString(R.string.merlindescription_key)))
        {
            return MediaPlayer.create(mContext, R.raw.merlindescription);
        }
        else if (res.equals(mContext.getString(R.string.percivaldescription_key)))
        {
            return MediaPlayer.create(mContext, R.raw.percivaldescription);
        }
        else if (res.equals(mContext.getString(R.string.loyaldescription_key)))
        {
            return MediaPlayer.create(mContext, R.raw.loyaldescription);
        }
        else if (res.equals(mContext.getString(R.string.assassindescription_key)))
        {
            return MediaPlayer.create(mContext, R.raw.assassindescription);
        }
        else if (res.equals(mContext.getString(R.string.morganadescription_key)))
        {
            return MediaPlayer.create(mContext, R.raw.morganadescription);
        }
        else if (res.equals(mContext.getString(R.string.mordreddescription_key)))
        {
            return MediaPlayer.create(mContext, R.raw.mordreddescription);
        }
        else if (res.equals(mContext.getString(R.string.oberondescription_key)))
        {
            return MediaPlayer.create(mContext, R.raw.oberondescription);
        }
        else if (res.equals(mContext.getString(R.string.miniondescription_key)))
        {
            return MediaPlayer.create(mContext, R.raw.miniondescription);
        }
        else if (res.equals(mContext.getString(R.string.liberaldescription_key)))
        {
            return MediaPlayer.create(mContext, R.raw.liberaldescription);
        }
        else if (res.equals(mContext.getString(R.string.hitlerdescription_key)))
        {
            return MediaPlayer.create(mContext, R.raw.hitlerdescription);
        }
        else if (res.equals(mContext.getString(R.string.fascistdescription_key)))
        {
            return MediaPlayer.create(mContext, R.raw.fascistdescription);
        }

        return null;
    }

    private Context mContext;

    private MediaPlayer mMediaPlayer;
    private int mMediaPlayerCurrentLength;
}

File: app/src/main/java/com/liweiyap/narradir/util/audio/CharacterDescriptionMediaPlayer.kt
package com.liweiyap.narradir.util.audio

import android.content.Context
import android.media.MediaPlayer

import com.liweiyap.narradir.R

class CharacterDescriptionMediaPlayer(context: Context): IMediaPlayerController {
    private var mContext: Context?
    private var mMediaPlayer: MediaPlayer? = null
    private var mMediaPlayerCurrentLength: Int = 0

    init {
        mContext = context
    }

    fun destroy() {
        mContext = null
        free()
    }

    override fun play(res: String?, volume: Float, listener: MediaPlayer.OnCompletionListener?) {
        mMediaPlayer?.stop()

        if (res == null) {
            return
        }

        try {
            mMediaPlayer = create(res)
            if (mMediaPlayer == null) {
                return
            }

            setVolume(volume)
            mMediaPlayer!!.setOnCompletionListener(listener)
            mMediaPlayer!!.start()
        } catch (e: IllegalStateException) {
            e.printStackTrace()
        }
    }

    override fun resume() {
        mMediaPlayer?.seekTo(mMediaPlayerCurrentLength)
        mMediaPlayer?.start()
    }

    override fun pause() {
        mMediaPlayer?.pause()
        mMediaPlayer?.let {
            mMediaPlayerCurrentLength = it.currentPosition
        }
    }

    override fun stop() {
        mMediaPlayer?.setOnCompletionListener(null)
        mMediaPlayer?.stop()
    }

    override fun free() {
        mMediaPlayer?.release()
        mMediaPlayer = null
    }

    override fun setVolume(volume: Float) {
        mMediaPlayer?.setVolume(volume, volume)
    }

    private fun create(res: String): MediaPlayer? {
        if (mContext == null) {
            return null
        }

        // does not look elegant but it's safe because RawRes ID integers are non-constant from Gradle version >4.
        // no need to call prepare(); create() does that for you (https://stackoverflow.com/a/59682667/12367873)
        return when (res) {
            mContext!!.getString(R.string.merlindescription_key) -> MediaPlayer.create(mContext, R.raw.merlindescription)
            mContext!!.getString(R.string.percivaldescription_key) -> MediaPlayer.create(mContext, R.raw.percivaldescription)
            mContext!!.getString(R.string.loyaldescription_key) -> MediaPlayer.create(mContext, R.raw.loyaldescription)
            mContext!!.getString(R.string.assassindescription_key) -> MediaPlayer.create(mContext, R.raw.assassindescription)
            mContext!!.getString(R.string.morganadescription_key) -> MediaPlayer.create(mContext, R.raw.morganadescription)
            mContext!!.getString(R.string.mordreddescription_key) -> MediaPlayer.create(mContext, R.raw.mordreddescription)
            mContext!!.getString(R.string.oberondescription_key) -> MediaPlayer.create(mContext, R.raw.oberondescription)
            mContext!!.getString(R.string.miniondescription_key) -> MediaPlayer.create(mContext, R.raw.miniondescription)
            mContext!!.getString(R.string.liberaldescription_key) -> MediaPlayer.create(mContext, R.raw.liberaldescription)
            mContext!!.getString(R.string.hitlerdescription_key) -> MediaPlayer.create(mContext, R.raw.hitlerdescription)
            mContext!!.getString(R.string.fascistdescription_key) -> MediaPlayer.create(mContext, R.raw.fascistdescription)
            else -> null
        }
    }
}